File: vocabulary-backend/src/main/java/com/languages/entities/Term.java
package com.languages.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import org.hibernate.annotations.*;

import javax.persistence.*;
import javax.persistence.CascadeType;
import javax.persistence.Entity;
import java.util.Collection;
import java.util.Date;

/**
 * Created by alexandrosfilios on 30/10/16.
 */
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "type")
@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.EXISTING_PROPERTY,
        property = "type",
        visible = true
)
@JsonSubTypes({
        @JsonSubTypes.Type(value = Noun.class, name = "NOUN"),
        @JsonSubTypes.Type(value = Phrase.class, name = "PHRASE"),
        @JsonSubTypes.Type(value = Verb.class, name = "VERB")
})
abstract public class Term {

    @Id
    @GeneratedValue
    private Long id;
    private String term;
    private String comments;
    @OneToMany(mappedBy = "term", cascade = CascadeType.ALL)
    private Collection<Definition> definitions;
    private Date createdOn;

    public abstract String getType();

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTerm() {
        return term;
    }

    public void setTerm(String term) {
        this.term = term;
    }

    public String getComments() {
        return comments;
    }

    public void setComments(String comments) {
        this.comments = comments;
    }

    public Collection<Definition> getDefinitions() {
        return definitions;
    }

    public void setDefinitions(Collection<Definition> definitions) {
        this.definitions = definitions;
        this.definitions.forEach(definition -> definition.setTerm(this));
    }

    public Date getCreatedOn() {
        return createdOn;
    }

    public void setCreatedOn(Date createdOn) {
        this.createdOn = createdOn;
    }
}

File: vocabulary-frontend/src/modules/add/core/entities/Term.ts
/**
 * Created by alexandrosfilios on 30/10/16.
 */

import {Definition} from "./Definition";
export enum Type {
    VERB = "VERB",
    NOUN = "NOUN",
    PHRASE = "PHRASE"
}
export enum Gender {
    MASCULINE = "MASCULINE",
    FEMININE = "FEMININE"
}
    
export class Term {
    public id: number;
    public type: Type;
    public term: string;
    public comments: string;
    public gender: Gender;
    public definitions: Array<Definition>;
    public createdOn: string;
    constructor(id?: number,
                type?: Type,
                term?: string,
                comments?: string,
                gender?: Gender,
                createdOn?: string,
                definitions?: Array<Definition>) {
        this.id = id || 0;
        this.type = type || Type.PHRASE;
        this.term = term || '';
        this.comments = comments || '';
        this.gender = gender || null;
        this.createdOn = createdOn || new Date();
        this.definitions = definitions || [];
    }
    public isEmpty(): boolean {
        return this.term.length > 0 && this.type;
    }
    public static getAllTypes(): Array<Type> {
        return [Type.NOUN, Type.PHRASE, Type.VERB];
    }
    public static getAllGenders(): Array<Gender> {
        return [Gender.FEMININE, Gender.MASCULINE];
    }
    public withExtraFields(): Term {
        this.definitions = this.definitions.concat(new Definition())
            .map(definition => definition.withExtraFields());
        return this;
    }
}
